File: frontend/src/lib/types/asset-query.ts
export type QueryFieldType =
	| 'string'
	| 'enum'
	| 'number'
	| 'bytes'
	| 'duration'
	| 'date'
	| 'ip'
	| 'text'
	| 'flag';

export interface QueryFieldSpec {
	name: string;
	type: QueryFieldType;
	group: string;
	description: string;
	example: string;
	aliases: string[];
	values: string[];
	facet: string | null;
	operators: string[];
	free_text: boolean;
	unit: string | null;
	dynamic_sub: string | null;
}

export interface QueryOperatorSpec {
	symbol: string;
	description: string;
}

export interface QueryFlagSpec {
	value: string;
	description: string;
}

export interface QueryExampleSpec {
	query: string;
	description: string;
	group: string;
	generic: boolean;
}

export interface QueryLead extends QueryExampleSpec {
	count: number;
	capped: boolean;
}

export type QueryStarter = QueryExampleSpec | QueryLead;

export interface QueryLeads {
	leads: QueryLead[];
	total: number;
	total_capped: boolean;
	filtered: boolean;
	computed: boolean;
}

export interface QueryGroupSpec {
	key: string;
	label: string;
	description: string;
}

export interface QueryGroup {
	value: string;
	label: string;
	count: number;
	query: string;
}

export interface QueryGroups {
	dimension: string;
	groups: QueryGroup[];
	total_groups: number;
	truncated: boolean;
	hosts: number;
	covered: number;
}

export interface QuerySchema {
	max_length: number;
	max_terms: number;
	groups: string[];
	example_groups: string[];
	group_dimensions: QueryGroupSpec[];
	fields: QueryFieldSpec[];
	operators: QueryOperatorSpec[];
	connectors: QueryOperatorSpec[];
	flags: QueryFlagSpec[];
	examples: QueryExampleSpec[];
}

export interface QueryError {
	message: string;
	hint: string | null;
	start: number;
	end: number;
}

export interface MatchEvidence {
	field: string;
	label: string;
	term: string;
	snippet: string | null;
}

export const EMPTY_QUERY_SCHEMA: QuerySchema = {
	max_length: 2000,
	max_terms: 8,
	groups: [],
	example_groups: [],
	group_dimensions: [],
	fields: [],
	operators: [],
	connectors: [],
	flags: [],
	examples: []
};

File: shared/models/asset_query.py
from pydantic import BaseModel, Field


class QueryFieldSpec(BaseModel):
    name: str
    type: str
    group: str
    description: str
    example: str
    aliases: list[str] = Field(default_factory=list)
    values: list[str] = Field(default_factory=list)
    facet: str | None = None
    operators: list[str] = Field(default_factory=list)
    free_text: bool = False
    unit: str | None = None
    dynamic_sub: str | None = None


class QueryOperatorSpec(BaseModel):
    symbol: str
    description: str


class QueryExampleSpec(BaseModel):
    query: str
    description: str
    group: str = ""
    generic: bool = False


class QueryLead(QueryExampleSpec):
    count: int = 0
    capped: bool = False


class QueryLeads(BaseModel):
    leads: list[QueryLead] = Field(default_factory=list)
    total: int = 0
    total_capped: bool = False
    filtered: bool = False
    computed: bool = False


class QueryGroupSpec(BaseModel):
    key: str
    label: str
    description: str


class QueryGroup(BaseModel):
    value: str
    label: str
    count: int
    query: str


class QueryGroups(BaseModel):
    dimension: str = ""
    groups: list[QueryGroup] = Field(default_factory=list)
    total_groups: int = 0
    truncated: bool = False
    hosts: int = 0
    covered: int = 0


class QueryFlagSpec(BaseModel):
    value: str
    description: str


class QuerySchema(BaseModel):
    max_length: int = 0
    max_terms: int = 0
    groups: list[str] = Field(default_factory=list)
    example_groups: list[str] = Field(default_factory=list)
    group_dimensions: list[QueryGroupSpec] = Field(default_factory=list)
    fields: list[QueryFieldSpec] = Field(default_factory=list)
    operators: list[QueryOperatorSpec] = Field(default_factory=list)
    connectors: list[QueryOperatorSpec] = Field(default_factory=list)
    flags: list[QueryFlagSpec] = Field(default_factory=list)
    examples: list[QueryExampleSpec] = Field(default_factory=list)


class QueryError(BaseModel):
    message: str
    hint: str | None = None
    start: int = 0
    end: int = 0


class MatchEvidence(BaseModel):
    field: str
    label: str
    term: str
    snippet: str | None = None
